### PR TITLE
Implement Admit and Validate stages for image policy admission

### PR DESCRIPTION
**Pull request by liggitt**
the ImagePolicy admission plugin must ensure the image spec on the incoming object still passes policy rules during the validation phase, in case a later admission plugin messed with a resolved/approved pull spec later in the mutation phase.

changes or policy failures result in rejection

### CODE REVIEW

File: pkg/image/admission/imagepolicy/imagepolicy_test.go [hunk 1]
```diff
@@ -99,29 +98,27 @@ func TestDefaultPolicy(t *testing.T) {
 	}
 
 	client := &imageclient.Clientset{}
-	imageResp := 0
-	// respond to images in this order: goodImage, badImage
 	client.AddReactor("get", "images", func(action clientgotesting.Action) (handled bool, ret runtime.Object, err error) {
-		image := goodImage
-		if imageResp%2 == 1 {
```
Comment (liggitt) [inline pkg/image/admission/imagepolicy/imagepolicy_test.go]: this switched from order-dependent responses to actually responding based on what was requested

File: pkg/image/admission/imagepolicy/imagepolicy_test.go [hunk 2]
```diff
@@ -99,29 +98,27 @@ func TestDefaultPolicy(t *testing.T) {
...
+		default:
+			return true, nil, kerrors.NewNotFound(imageapi.Resource("images"), name)
 		}
-		imageResp += 1
-		return true, image, nil
 	})
-	tagResp := 0
-	// respond to imagestreamtags in this order: notFoundTag, goodTag, badTag
 	client.AddReactor("get", "imagestreamtags", func(action clientgotesting.Action) (handled bool, ret runtime.Object, err error) {
-		if tagResp%3 == 0 {
```
Comment (liggitt) [inline pkg/image/admission/imagepolicy/imagepolicy_test.go]: same here

File: pkg/image/admission/imagepolicy/imagepolicy_test.go [hunk 3]
```diff
@@ -429,12 +480,20 @@ func TestAdmissionResolution(t *testing.T) {
...
+		pod.Spec.Containers[1].Image != "myregistry.com/mysql/mysql@sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4" {
+		t.Errorf("unexpected image: %#v", pod)
+	}
 }
 
 func TestAdmissionResolveImages(t *testing.T) {
 	image1 := &imageapi.Image{
 		ObjectMeta:           metav1.ObjectMeta{Name: "sha256:0000000000000000000000000000000000000000000000000000000000000001"},
-		DockerImageReference: "integrated.registry/image1/image1:latest",
+		DockerImageReference: "integrated.registry/image1/image1@sha256:0000000000000000000000000000000000000000000000000000000000000001",
```
Comment (liggitt) [inline pkg/image/admission/imagepolicy/imagepolicy_test.go]: this was bad test data... in real life, all `Image` objects have a `DockerImageReference` containing a sha, not a tag... putting a tag here meant the build referencing a specific sha got resolved to a tag (which is wrong), and then tried to re-resolve to a sha during the validation stage

**Comment by liggitt** [comment]
@deads2k for admission
@bparees and @smarterclayton for image policy

**Comment by liggitt** [comment]
fyi @sjenning @frobware, relevant to the image qualification admission work

**Comment by deads2k** [comment]
Admission change is shockingly small.  That bit lgtm.

File: pkg/image/admission/imagepolicy/imagepolicy_test.go [hunk 4]
```diff
@@ -406,6 +454,9 @@ func TestAdmissionResolution(t *testing.T) {
 	if err := p.Admit(failingAttrs); err == nil {
 		t.Fatal(err)
 	}
+	if err := p.Validate(failingAttrs); err == nil {
+		t.Fatal(err)
```
Comment (bparees) [inline pkg/image/admission/imagepolicy/imagepolicy_test.go]: needs a better log statement, this is going to fail the test with err reason "nil"

Comment (bparees) [reply pkg/image/admission/imagepolicy/imagepolicy_test.go]: (copied from the one above which should also be fixed...)

File: pkg/image/admission/imagepolicy/imagepolicy_test.go [hunk 5]
```diff
@@ -306,9 +345,13 @@ func TestDefaultPolicy(t *testing.T) {
 		"default", "pod1", schema.GroupVersionResource{Version: "v1", Resource: "pods"},
 		"", admission.Create, nil,
 	)
+	t.Log("here")
```
Comment (bparees) [inline pkg/image/admission/imagepolicy/imagepolicy_test.go]: delete

Comment (liggitt) [reply pkg/image/admission/imagepolicy/imagepolicy_test.go]: done

File: pkg/image/admission/imagepolicy/imagepolicy_test.go (hunk 4 above)
Comment (liggitt) [reply pkg/image/admission/imagepolicy/imagepolicy_test.go]: it'll include the line number, which gets you to the failed case... I'd rather keep the changes obviously additive than start changing the existing tests

**Comment by liggitt** [comment]
> I don't see something that actually test the "mutation disallowed" logic in validate?

Added a mutation test https://github.com/openshift/origin/pull/18087/files#diff-95d75acb7f2d1edc75054d90deae1a81R491

**Comment by bparees** [comment]
lgtm
